clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -209,7 +209,10 @@ public RecordAppendResult append(TopicPartition tp,
                     return appendResult;
                 }
 
-                MemoryRecordsBuilder recordsBuilder = recordsBuilder(buffer, maxUsableMagic);
+                // We have to pass `size` because the buffer returned from `free.allocate` could, in theory, be larger
+                // than the requested size if `deallocate(ByteBuffer buffer, int size)` is invoked with a buffer larger
+                // than the `size` parameter
```

Comment:
Hmm, not sure if the comment is accurate. It seems that the returned buffer should always be of size. Buffers with size other than batchSize are never reused after being freed. Size can be larger that batchSize.